# Machine learning HomeWork 1

## First problem

In [100]:
sentences = open('sentences.txt', 'r') # open file
s = sentences.read().lower().splitlines() #read, tolower and split sentences
s

['in comparison to dogs, cats have not undergone major changes during the domestication process.',
 'as cat simply catenates streams of bytes, it can be also used to concatenate binary files, where it will just concatenate sequence of bytes.',
 'a common interactive use of cat for a single file is to output the content of a file to standard output.',
 'cats can hear sounds too faint or too high in frequency for human ears, such as those made by mice and other small animals.',
 'in one, people deliberately tamed cats in a process of artificial selection, as they were useful predators of vermin.',
 'the domesticated cat and its closest wild ancestor are both diploid organisms that possess 38 chromosomes and roughly 20,000 genes.',
 'domestic cats are similar in size to the other members of the genus felis, typically weighing between 4 and 5 kg (8.8 and 11.0 lb).',
 'however, if the output is piped or redirected, cat is unnecessary.',
 'cat with one named file is safer where human error i

In [188]:
import re
dataArray = [None] * len(s) # declare result list
for i in range(len(s)): 
    tokens = re.split('[^a-z]', s[i]) # tokenizes sentences one at a time
    filter_object = filter(lambda x: x != "", tokens)  # filters empty strings
    tokens = list(filter_object) # reassigns filtered list
    dataArray[i] = tokens # assigning tokenized arrays
print(dataArray)

[['in', 'comparison', 'to', 'dogs', 'cats', 'have', 'not', 'undergone', 'major', 'changes', 'during', 'the', 'domestication', 'process'], ['as', 'cat', 'simply', 'catenates', 'streams', 'of', 'bytes', 'it', 'can', 'be', 'also', 'used', 'to', 'concatenate', 'binary', 'files', 'where', 'it', 'will', 'just', 'concatenate', 'sequence', 'of', 'bytes'], ['a', 'common', 'interactive', 'use', 'of', 'cat', 'for', 'a', 'single', 'file', 'is', 'to', 'output', 'the', 'content', 'of', 'a', 'file', 'to', 'standard', 'output'], ['cats', 'can', 'hear', 'sounds', 'too', 'faint', 'or', 'too', 'high', 'in', 'frequency', 'for', 'human', 'ears', 'such', 'as', 'those', 'made', 'by', 'mice', 'and', 'other', 'small', 'animals'], ['in', 'one', 'people', 'deliberately', 'tamed', 'cats', 'in', 'a', 'process', 'of', 'artificial', 'selection', 'as', 'they', 'were', 'useful', 'predators', 'of', 'vermin'], ['the', 'domesticated', 'cat', 'and', 'its', 'closest', 'wild', 'ancestor', 'are', 'both', 'diploid', 'organism

In [104]:
wordsDict = {} # declare dictionary
cnt = 0 # declare index counter
for i in range(len(dataArray)):
    for j in dataArray[i]:
        if j not in wordsDict: # ensuring indeces are in order
            wordsDict[j] = cnt # assign index
            cnt += 1 #increment counter
               
print(wordsDict)              

{'in': 0, 'comparison': 1, 'to': 2, 'dogs': 3, 'cats': 4, 'have': 5, 'not': 6, 'undergone': 7, 'major': 8, 'changes': 9, 'during': 10, 'the': 11, 'domestication': 12, 'process': 13, 'as': 14, 'cat': 15, 'simply': 16, 'catenates': 17, 'streams': 18, 'of': 19, 'bytes': 20, 'it': 21, 'can': 22, 'be': 23, 'also': 24, 'used': 25, 'concatenate': 26, 'binary': 27, 'files': 28, 'where': 29, 'will': 30, 'just': 31, 'sequence': 32, 'a': 33, 'common': 34, 'interactive': 35, 'use': 36, 'for': 37, 'single': 38, 'file': 39, 'is': 40, 'output': 41, 'content': 42, 'standard': 43, 'hear': 44, 'sounds': 45, 'too': 46, 'faint': 47, 'or': 48, 'high': 49, 'frequency': 50, 'human': 51, 'ears': 52, 'such': 53, 'those': 54, 'made': 55, 'by': 56, 'mice': 57, 'and': 58, 'other': 59, 'small': 60, 'animals': 61, 'one': 62, 'people': 63, 'deliberately': 64, 'tamed': 65, 'artificial': 66, 'selection': 67, 'they': 68, 'were': 69, 'useful': 70, 'predators': 71, 'vermin': 72, 'domesticated': 73, 'its': 74, 'closest': 

In [190]:
import numpy as np
 # declare matrix with height and weight corresponding to amount of sentences and unique word amount
arr = np.zeros((len(dataArray),wordsDict.get(list(wordsDict)[-1]) + 1))

In [191]:
for i in dataArray: # in every sentence
    for j in wordsDict: # check amount of words occurings
        arr[dataArray.index(i), wordsDict.get(j)] = i.count(j)
arr

array([[1., 1., 1., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       [0., 0., 2., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [1., 0., 1., ..., 0., 0., 0.],
       [0., 0., 1., ..., 1., 1., 1.]])

In [143]:
# see results with library function to compare
from scipy.spatial.distance import cdist # cosine distance funciton import

cdist(arr[0:1], arr[:], metric='cosine')

array([[0.     , 0.95275, 0.86447, 0.89517, 0.77709, 0.94024, 0.73274,
        0.92588, 0.88427, 0.90551, 0.83282, 0.88048, 0.83964, 0.87036,
        0.87401, 0.94427, 0.84064, 0.95664, 0.94427, 0.88854, 0.84276,
        0.82504]])

In [194]:
# Write Cosine distance evaluator for matrices
from numpy.linalg.linalg import norm
from numpy import set_printoptions

# dot products of rows against themselves
dotProduct = arr.dot(arr.T);       

# denominator of a formula, multiplication of norms
denominator = a([norm(arr, axis=1)]) * a([norm(arr, axis=1)]).T; 

# evaluate the cosine distance
cosineDistance = 1 - (dotProduct / denominator)

# set precision to 2, for comparison convenience
set_printoptions(precision=2, suppress=True)

print(cosineDistance[0])

[0.   0.95 0.86 0.9  0.78 0.94 0.73 0.93 0.88 0.91 0.83 0.88 0.84 0.87
 0.87 0.94 0.84 0.96 0.94 0.89 0.84 0.83]


### 0.96,0.95,0.94(three times)


# As we can see, 0.96 (index 17) and 0.95(index 1) are the closest.
### So Answer is 
### CosDistance of 0.96 sentence 18(index + 1) and 0.95 sentence 2(index + 1), are the closest

### But all 3 of the sentences(including first) have different subjects , of the next three 0.94 cos distance values only one has same subject, 6th sentence